vertx-web/src/test/java/io/vertx/ext/web/RouterTest.java

```diff
@@ -1855,4 +1856,35 @@ public void testGetWithPlusPath2() throws Exception {
     });
     testRequest(HttpMethod.GET, "/some+path?q1=some+query", 200, "foo");
   }
+
+  @Test
+  public void testMethodNotAllowed() throws Exception {
+    router.get("/abc").handler(context -> {
+      fail("Should fail!");
+    });
+
+    testRequest(HttpMethod.POST, "/abc", null, resp -> {
+      assertEquals(405, resp.statusCode());
+      MultiMap headers = resp.headers();
+      assertTrue(headers.contains("Allow"));
+      assertEquals("GET", headers.get("Allow"));
+    }, 405, "Method Not Allowed", null);
+  }
+
+  @Test
+  public void testMethodNotAllowed2() throws Exception {
+    router.clear();
+    router.route().path("/abc").method(HttpMethod.GET).method(HttpMethod.PUT).handler(context -> {
+      fail("Should fail!");
+    });
```

Comment:
That test won't work and making it work requires a huge refactoring to the router class. The issue is that `Router` was built with the concept of independent `Route`'s. So each `Route` is not aware of its surroundings. In the case you describe the second `Route` does not have a context of what happened before, and the first one does not know what would happen next.

It is a limitation that I want to discuss during the F2F meeting next week.